core/src/main/scala/org/apache/spark/scheduler/SchedulerBackend.scala

```diff
@@ -41,4 +41,11 @@ private[spark] trait SchedulerBackend {
    */
   def applicationId(): String = appId
 
+  /**
+   * Get an application ID associated with the job.
+   *
+   * @return An application attempt id
+   */
+  def applicationAttemptId(): String = ""
```

Comment:
I think it would be cleaner if you used an `Option[String]` with a default of `None`.  As I was reading the rest of the code I was wondering if every framework would be forced to provide some dummy attempt id, or if would be null, etc.  `Option` would make that more explicit.  Plus it seems you use an option later in the listener event for json in any case.